parlai/tasks/dailydialog/agents.py

```diff
@@ -88,3 +88,43 @@ def share(self):
         shared = super().share()
         shared['data'] = self.data
         return shared
+
+
+class NoSilenceTeacher(Convai2Teacher):
+    # Same as the default teacher, but it doesn't contain __SILENCE__ entries.
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+
+        # Calculate the correct number of examples.
+        self.num_exs = sum(len(d['dialogue']) - 1 for d in self.data)
+
+    def get(self, episode_idx, entry_idx=0):
+        # Sometimes we're speaker 1 and sometimes we're speaker 2
+        speaker_id = episode_idx % 2
+        full_eps = self.data[episode_idx // 2]
+        entries = full_eps['dialogue']
+
+        # If we are first speaker, then the first utterance is skipped.
+        # If dialog is only 2 turns, then we can't shift it.
+        if len(entries) > 2:
```

Comment:
Honestly it feels like we should be able to do this without special casing using correct math. If a conversation only has 2 turns, then we shouldn't have it included in num_episodes. This makes all of this much more nuanced :(

Reply:
I will think about it, it might be easier than I originally thought